src/main/java/net/minecraftforge/registries/DeferredRegister.java

```diff
@@ -80,13 +107,43 @@ public DeferredRegister(IForgeRegistry<T> reg, String modid)
         Objects.requireNonNull(name);
         Objects.requireNonNull(sup);
         final ResourceLocation key = new ResourceLocation(modid, name);
-        RegistryObject<I> ret = RegistryObject.of(key, this.type);
+        RegistryObject<I> ret;
+        try
+        {
+            ret = RegistryObject.of(key, this.type);
+        }
+        catch (IllegalArgumentException e)
+        {
+            LOGGER.error("Errored while trying to register an entry. This likely happened due to registering for a custom registry, use lazyRegister instead.");
+            throw new IllegalArgumentException("Null registry used in DeferredRegister");
+        }
         if (entries.putIfAbsent((RegistryObject<T>) ret, () -> sup.get().setRegistryName(key)) != null) {
             throw new IllegalArgumentException("Duplicate registration " + name);
         }
         return ret;
     }
 
+    /**
+     * Adds a new supplier to the list of entries to be registered, and returns a {@link LazyRegistryObject}
+     * that will be populated with the created entry automatically and lazily.
+     *
+     * @param name The new entry's name, it will automatically have the modid prefixed.
+     * @param sup A factory for the new entry, it should return a new instance every time it is called.
+     * @return A {@link LazyRegistryObject} that will be lazily updated with when the entries in the registry change.
+     */
+    public <I extends T> LazyRegistryObject<I> lazyRegister(final String name, final Supplier<? extends I> sup)
+    {
+        Objects.requireNonNull(name);
+        Objects.requireNonNull(sup);
+        final ResourceLocation key = new ResourceLocation(modid, name);
+        LazyRegistryObject<I> lazyRet = new LazyRegistryObject<>(()->RegistryObject.of(key, this.type));
```

Comment:
Instead of giving it `this.type` in the supplier for the lazy registry object, shouldn't it be `getFromSupplier()` to have slightly less reliance on making sure that `getFromSupplier()` was called before the supplier is ran. I believe this may be the case if `getEntries()` is called after the new registry event but before the entries are actually added in `addEntries` (Not that I can think of a case for why someone would call that in that stage)

Reply:
I can test that. I put this.type directly to be sure to get an error since that would mean something went wrong compared to the intention

Reply:
so yeah... it indeeds prevent a crash if you getEntries() in between the NewRegistry Event and the Registry<Block> event. so changed it. but would be a weird occurence since you have a list of registry objects that are effectively empty